raft.go

```diff
@@ -252,6 +360,60 @@ func (l *Leader) SendHeartbeat(handler RequestHandler) {
 		if i == l.cfg.CurrentId {
 			continue
 		}
-		handler.HandleAppendEntryRequest(AppendEntryRequest{ReceiverId: i, Id: l.cfg.CurrentId, Term: l.term})
+
+		entries := make([]LogEntry, 0, 1)
+		if l.nextIndex[i] < len(l.log) {
+			entries = append(entries, l.log[l.nextIndex[i]])
+		}
+		request := AppendEntryRequest{
+			ReceiverId:   i,
+			Id:           l.cfg.CurrentId,
+			Term:         l.term,
+			Entries:      entries,
+			LeaderCommit: l.commitIndex,
+			PrevLogIndex: l.nextIndex[i] - 1,
+			PrevLogTerm:  l.log[l.nextIndex[i]-1].Term,
+		}
+		handler.HandleAppendEntryRequest(request)
+	}
+}
+
+func (l *Leader) ExecuteCommand(cmd Command) {
+	l.log = append(l.log, LogEntry{Term: l.Term(), Command: cmd})
+}
+
+func (l *Leader) HandleAppendEntryResponse(r AppendEntryResponse) *Follower {
+	if l.Term() != r.RequestInTerm {
+		return nil
+	}
+	if l.Term() < r.Term {
+		f := MakeFollower(l.cfg, r.Term, l.log, l.commitIndex)
+		return &f
+	}
+	if r.Success {
+		l.matchIndex[r.ReceiverId] = r.MatchIndex
+		l.nextIndex[r.ReceiverId] = r.MatchIndex + 1
+		l.updateCommitIndex()
+	} else {
+		l.nextIndex[r.ReceiverId]--
+	}
+
+	return nil
+}
+
+func (l *Leader) updateCommitIndex() {
+	matches := make([]int, len(l.cfg.Cluster))
+	for i, _ := range matches {
+		if i == l.cfg.CurrentId {
+			matches[i] = len(l.log) - 1
+			continue
+		}
+		matches[i] = l.matchIndex[i]
+	}
+
+	sort.Ints(matches)
```

Comment:
this one might be slow I presume? can we do it better?

Reply:
I thought about it a lot. There is better solution - keep in memory sorted list and swap items on demand. 
But it algorithm is more cumbersome.
What do you think, should I try to implement it or maybe you have idea of another better solution?